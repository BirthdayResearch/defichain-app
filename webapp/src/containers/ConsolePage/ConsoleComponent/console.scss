@import '~bootstrap/scss/functions';
@import '~bootstrap/scss/variables';
@import '~bootstrap/scss/mixins';
@import '../../../scss/variables';
@import '../../../scss/bootstrap/variables';

div.react-console-container {
  font-size: 0.875em;
  font-family: $font-family-monospace;
  min-height: calc(100vh - 112px);
  box-sizing: border-box;
  white-space: pre;
  padding: 2rem;
  cursor: text;
}

// div.react-console-prompt-box {

// }

span.react-console-prompt-label {
  font-weight: $font-weight-semibold;
  color: $secondary;
  line-height: 24px;
  text-align: right;
  display: inline-block;
}

.react-console-prompt-label::before {
  content: '\232A';
  padding-right: 10px;
}

span.react-console-prompt {
  color: $dfi-pink;
  line-height: 1.5rem;
  display: inline-block;
}

div.react-console-focus span.react-console-cursor {
  background: $dfi-pink;
}

span.react-console-cursor {
  display: inline-block;
  content: '';
  width: 8px;
  height: 1.5rem;
  vertical-align: bottom;
}

div.react-console-nofocus span.react-console-cursor {
  background: none;
}

div.react-console-focus span.react-console-cursor-idle {
  animation: blink 1s infinite;
}

div.react-console-welcome {
  font-weight: $font-weight-medium;
}

@keyframes blink {
  0% {
    background-color: $dfi-pink;
  }

  25% {
    background-color: $dfi-pink;
  }

  26% {
    background-color: transparent;
  }

  75% {
    background-color: transparent;
  }

  76% {
    background-color: $dfi-pink;
  }
}

div.react-console-message {
  word-break: break-all;
  white-space: pre-wrap;
  line-height: 1.5rem;
}

div.react-console-prompt-box {
  span {
    word-break: break-all;
    white-space: pre-line;
    color: $body-color;
  }
}

/*# sourceMappingURL=react-console.css.map*/
