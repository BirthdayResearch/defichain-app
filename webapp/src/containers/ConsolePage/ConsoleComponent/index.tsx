import React, { useEffect, useState } from 'react';
import Console from 'react-console-component';
import cloneDeep from 'lodash/cloneDeep';
import isEmpty from 'lodash/isEmpty';
import { connect, useSelector } from 'react-redux';
import {
  fetchDataForQueryRequest,
  resetDataForQuery,
  storeCliLog,
} from '../reducer';
import {
  CONSOLE_PROMPT_LABEL,
  BITCOIN_CLI_REGEX,
  DEFI_CLI_TEXT,
} from '../../../constants';
import './console.scss';
import { WalletMap } from '@defi_types/walletMap';
import { RootState } from '../../../app/rootTypes';

interface EchoConsoleProps {
  fetchDataForQueryRequest: (query: string) => void;
  resetDataForQuery: () => void;
  storeCliLog: (logState: any) => void;
  cliLog: any;
  isLoading: boolean;
  result: string | number | any[];
  isError: string;
  walletMap: WalletMap;
}

const EchoConsole: React.FunctionComponent<EchoConsoleProps> = (
  props: EchoConsoleProps
) => {
  const {
    fetchDataForQueryRequest,
    resetDataForQuery,
    storeCliLog,
    isLoading,
    result,
    isError,
    cliLog,
    walletMap,
  } = props;
  const { nodeVersion } = useSelector((state: RootState) => state.app);
  const [showConsoleResults, setShowConsoleResults] = useState(false);
  let currentRef: Console;
  const echo = (text: string) => {
    if (!!currentRef) {
      if (text === 'clear') {
        currentRef.setState({
          acceptInput: true,
          log: [],
        });
        return storeCliLog(
          cloneDeep({
            history: currentRef.state.history,
          })
        );
      }
      if (!showConsoleResults) setShowConsoleResults(true);
      return fetchDataForQueryRequest(text);
    }
  };
  useEffect(() => {
    resetDataForQuery();
    if (currentRef.state && !isEmpty(cliLog)) {
      const updatedState = {
        log: [],
        history: [],
      };

      if (Array.isArray(cliLog.log) && cliLog.log.length > 0) {
        updatedState.log = cloneDeep(cliLog.log);
      }

      if (Array.isArray(cliLog.history) && cliLog.history.length > 0) {
        updatedState.history = cloneDeep(cliLog.history);
      }

      currentRef.setState(updatedState);
      if (cliLog.scrollSemaphore) {
        currentRef.scrollSemaphore = cliLog.scrollSemaphore;
      }
    }
  }, []);

  const printToConsole = (text: string) => {
    if (!!currentRef && !!text && showConsoleResults) {
      const updatedText = text.replace(BITCOIN_CLI_REGEX, DEFI_CLI_TEXT);
      currentRef.log(updatedText);
      currentRef.return();
      const { log, history } = currentRef.state;
      const updatedState = cloneDeep({
        log,
        history,
        scrollSemaphore: currentRef.scrollSemaphore,
      });
      storeCliLog(updatedState);
    }
  };

  const onPasteHandler = (e) => {
    if (!!currentRef) {
      currentRef.paste(e);
    }
  };

  useEffect(() => {
    let toPrintData;
    if (!isLoading) {
      if (!isError) {
        if (result !== undefined) {
          toPrintData = JSON.stringify(result, null, 2);
          if (typeof result === 'string') {
            toPrintData = result;
          }
        }
      } else {
        toPrintData = isError;
      }
      printToConsole(toPrintData);
    }
  }, [isLoading, result, isError]);

  const onFocusHandler = () => {
    return !!currentRef && currentRef.focus();
  };

  const onScrollHandler = () => {
    return !!currentRef && currentRef.scrollIfBottom();
  };

  const nv = nodeVersion ?? (walletMap?.nodeVersion ? `${walletMap.nodeVersion}` : '')

  return (
    <div
      onPaste={onPasteHandler}
      onFocus={onFocusHandler}
      onScroll={onScrollHandler}
    >
      <Console
        ref={(ref: Console) => (currentRef = ref)}
        handler={echo}
        promptLabel=''
        welcomeMessage={`${CONSOLE_PROMPT_LABEL}${
          nv
        }`}
        autofocus={true}
      />
    </div>
  );
};

const mapStateToProps = (state) => {
  const {
    cli: { isLoading, result, isError, cliLog },
    wallet: { walletMap },
  } = state;
  return {
    isLoading,
    result,
    isError,
    cliLog,
    walletMap,
  };
};
const mapDispatchToProps = {
  fetchDataForQueryRequest: (query: string) =>
    fetchDataForQueryRequest({ query }),
  resetDataForQuery,
  storeCliLog: (logState: any) => storeCliLog(logState),
};

export default connect(mapStateToProps, mapDispatchToProps)(EchoConsole);
