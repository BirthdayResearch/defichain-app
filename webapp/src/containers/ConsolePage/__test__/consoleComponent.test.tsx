import React from 'react';
import ConsoleComponent from '../ConsoleComponent';
import { Provider } from 'react-redux';
import store from '../../../app/rootStore';
import Enzyme from '../../../utils/testUtils/enzyme';

describe('Console component', () => {
  it('snapshot test of console component', () => {
    window.getSelection = jest.fn().mockImplementation(() => {
      return {
        toString: jest.fn(),
      };
    });
    const wrapper = Enzyme.mount(
      <Provider store={store}>
        <ConsoleComponent />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
