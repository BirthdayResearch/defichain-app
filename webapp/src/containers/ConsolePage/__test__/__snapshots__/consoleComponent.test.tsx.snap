// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Console component snapshot test of console component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(EchoConsole)>
    <EchoConsole
      cliLog={Object {}}
      fetchDataForQueryRequest={[Function]}
      isError=""
      isLoading={false}
      resetDataForQuery={[Function]}
      result={Object {}}
      storeCliLog={[Function]}
    >
      <div
        onFocus={[Function]}
        onPaste={[Function]}
        onScroll={[Function]}
      >
        <default_1
          autofocus={true}
          cancel={[Function]}
          continue={[Function]}
          handler={[Function]}
          promptLabel=""
          welcomeMessage="defi-wallet_v2.5.1"
        >
          <div
            className="react-console-container react-console-focus"
            onClick={[Function]}
          >
            <div
              className="react-console-message react-console-welcome"
            >
              defi-wallet_v2.5.1
            </div>
            <ConsolePrompt
              argument={null}
              label=""
              point={0}
              value=""
            >
              <div
                className="react-console-prompt-box"
              >
                <span
                  className="react-console-prompt-label"
                />
                <span
                  className="react-console-prompt-argument"
                />
                <span
                  className="react-console-prompt"
                >
                  <span
                    className="react-console-cursor"
                    key="cursor"
                  >

                  </span>
                </span>
              </div>
            </ConsolePrompt>
            <div
              style={
                Object {
                  "height": 1,
                  "overflow": "hidden",
                  "width": 1,
                }
              }
            >
              <textarea
                autoCapitalize="off"
                autoComplete="off"
                autoCorrect="off"
                className="react-console-typer"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                onPaste={[Function]}
                spellCheck="false"
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "border": "none",
                    "color": "transparent",
                    "outline": "none",
                    "overflow": "hidden",
                    "resize": "none",
                  }
                }
              />
            </div>
            <div>

            </div>
          </div>
        </default_1>
      </div>
    </EchoConsole>
  </Connect(EchoConsole)>
</Provider>
`;
